File: pyorion/api/dialog.py
# Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
# SPDX-License-Identifier: Apache-2.0
# SPDX-License-Identifier: MIT

"""Dialog API - File pickers and message dialogs.

Wrapper around the Rust event loop backend to show dialogs,
pick files or directories, and display messages.
"""

from typing import Any, Literal, Optional

from pyorion.runtime.runtime_handle import event_register


MessageLevel = Literal["info", "warning", "error"]


class DialogAPI:
    """Asynchronous API wrapper for dialog operations."""

    def __init__(self) -> None:
        """Initialize a new :class:`DialogAPI` instance."""

    async def show_message(
        self,
        title: str,
        content: Optional[str] = None,
        level: Optional[MessageLevel] = "info",
    ) -> None:
        """Show a message dialog.

        :param title: The title of the dialog window.
        :type title: str
        :param content: The main message text to display.
        :type content: Optional[str]
        :param level: Message level (``"info"``, ``"warning"``, or ``"error"``).
        :type level: Optional[MessageLevel]

        :return: Nothing.
        :rtype: None
        """
        args = [title, content, level]
        return await event_register("dialog.showMessage", args=args)

    async def pick_file(
        self,
        filters: Optional[list[str]] = None,
        start_dir: Optional[str] = None,
    ) -> dict[str, Any]:
        """Open a file picker dialog for a single file.

        :param filters: List of file type filters (e.g. ``["*.txt", "*.jpg"]``).
        :type filters: Optional[list[str]]
        :param start_dir: The starting directory path.
        :type start_dir: Optional[str]

        :return: Information about the selected file.
        :rtype: dict[str, Any]
        """
        args = [filters, start_dir]
        return await event_register("dialog.pickFile", args=args)

    async def pick_files(
        self,
        filters: Optional[list[str]] = None,
        start_dir: Optional[str] = None,
    ) -> dict[str, Any]:
        """Open a file picker dialog for multiple files.

        :param filters: List of file type filters.
        :type filters: Optional[list[str]]
        :param start_dir: The starting directory path.
        :type start_dir: Optional[str]

        :return: Information about the selected files.
        :rtype: dict[str, Any]
        """
        args = [filters, start_dir]
        return await event_register("dialog.pickFiles", args=args)

    async def pick_dir(
        self,
        start_dir: Optional[str] = None,
    ) -> dict[str, Any]:
        """Open a directory picker dialog for a single directory.

        :param start_dir: The starting directory path.
        :type start_dir: Optional[str]

        :return: Information about the selected directory.
        :rtype: dict[str, Any]
        """
        args = [start_dir]
        return await event_register("dialog.pickDir", args=args)

    async def pick_dirs(
        self,
        start_dir: Optional[str] = None,
    ) -> dict[str, Any]:
        """Open a directory picker dialog for multiple directories.

        :param start_dir: The starting directory path.
        :type start_dir: Optional[str]

        :return: Information about the selected directories.
        :rtype: dict[str, Any]
        """
        args = [start_dir]
        return await event_register("dialog.pickDirs", args=args)

    async def save_file(
        self,
        filters: Optional[list[str]] = None,
        start_dir: Optional[str] = None,
    ) -> dict[str, Any]:
        """Open a file save dialog.

        :param filters: List of file type filters.
        :type filters: Optional[list[str]]
        :param start_dir: The starting directory path.
        :type start_dir: Optional[str]

        :return: Information about the saved file.
        :rtype: dict[str, Any]
        """
        args = [filters, start_dir]
        return await event_register("dialog.saveFile", args=args)

File: src/api/dialog.rs
// Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use anyhow::Result;
use pyorion_macros::api;

use crate::api_manager::ApiManager;
use serde::Deserialize;
use serde_json::{json, Value};
use tao::window::Window;

pub fn dialog_api(_api_manager: &mut ApiManager) {
    _api_manager.register_api("dialog.showMessage", show_message);
    _api_manager.register_api("dialog.pickFile", pick_file);
    _api_manager.register_api("dialog.pickFiles", pick_files);
    _api_manager.register_api("dialog.pickDir", pick_dir);
    _api_manager.register_api("dialog.pickDirs", pick_dirs);
    _api_manager.register_api("dialog.saveFile", save_file);
}

#[derive(Deserialize)]
enum MessageLevel {
    #[serde(rename = "info")]
    Info,
    #[serde(rename = "warning")]
    Warning,
    #[serde(rename = "error")]
    Error,
}

#[api]
fn show_message(title: String, content: Option<String>, level: Option<MessageLevel>) -> Result<()> {
    let parent = app.app_context()?.get_window()?;
    let content = content.unwrap_or_default();
    let level = level.unwrap_or(MessageLevel::Info);

    rfd::MessageDialog::new()
        .set_title(&title)
        .set_description(&content)
        .set_parent(&parent)
        .set_level(match level {
            MessageLevel::Info => rfd::MessageLevel::Info,
            MessageLevel::Warning => rfd::MessageLevel::Warning,
            MessageLevel::Error => rfd::MessageLevel::Error,
        })
        .show();

    Ok(())
}

fn _create_dialog(
    parent: &Window,
    filters: Option<Vec<String>>,
    start_dir: Option<String>,
) -> rfd::FileDialog {
    let mut dialog = rfd::FileDialog::new();
    if let Some(extensions) = filters {
        if !extensions.is_empty() {
            let extensions = extensions.iter().map(|s| s.as_str()).collect::<Vec<&str>>();
            dialog = dialog.add_filter("pick", &extensions);
        }
    }
    if let Some(dir) = start_dir {
        dialog = dialog.set_directory(dir);
    }
    dialog.set_parent(parent)
}

// In allen relevanten Funktionen (ohne & vor parent):

#[api]
fn pick_file(filters: Option<Vec<String>>, start_dir: Option<String>) -> Result<Value> {
    let parent = app.app_context()?.get_window()?;
    let dialog = _create_dialog(&parent, filters, start_dir);

    match dialog.pick_file() {
        Some(file) => Ok(json!(file)),
        None => Ok(json!(null)),
    }
}

#[api]
fn pick_files(filters: Option<Vec<String>>, start_dir: Option<String>) -> Result<Value> {
    let parent = app.app_context()?.get_window()?;
    let dialog = _create_dialog(&parent, filters, start_dir);

    match dialog.pick_files() {
        Some(files) => Ok(json!(files)),
        None => Ok(json!(null)),
    }
}

#[api]
fn pick_dir(start_dir: Option<String>) -> Result<Value> {
    let parent = app.app_context()?.get_window()?;
    let dialog = _create_dialog(&parent, None, start_dir);

    match dialog.pick_folder() {
        Some(dir) => Ok(json!(dir)),
        None => Ok(json!(null)),
    }
}

#[api]
fn pick_dirs(start_dir: Option<String>) -> Result<Value> {
    let parent = app.app_context()?.get_window()?;
    let dialog = _create_dialog(&parent, None, start_dir);

    match dialog.pick_folders() {
        Some(dirs) => Ok(json!(dirs)),
        None => Ok(json!(null)),
    }
}

#[api]
fn save_file(filters: Option<Vec<String>>, start_dir: Option<String>) -> Result<Value> {
    let parent = app.app_context()?.get_window()?;
    let dialog = _create_dialog(&parent, filters, start_dir);

    match dialog.save_file() {
        Some(file) => Ok(json!(file)),
        None => Ok(json!(null)),
    }
}
